components/eventing-controller/controllers/subscription/nats/integration_test.go

```diff
@@ -0,0 +1,837 @@
+	testID                    int
+	cfg                       *rest.Config
+	k8sClient                 client.Client
+	testEnv                   *envtest.Environment
+	natsServer                *natsserver.Server
+	reconciler                *natsreconciler.Reconciler
+	natsBackend               *handlers.Nats
+	defaultSubscriptionConfig env.DefaultSubscriptionConfig
+	subscriberSvc             *v1.Service
+	cancel                    context.CancelFunc
+}
+
+type expect struct {
+	k8sSubscription  []gomegatypes.GomegaMatcher
+	k8sEvents        []v1.Event
+	natsSubscription []gomegatypes.GomegaMatcher
+}
+
+//TestUnavailableNATSServer tests if a subscription is reconciled properly when the NATS backend is unavailable.
+func TestUnavailableNATSServer(t *testing.T) {
+	natsPort := 4220
+	ctx := context.Background()
+	g := gomega.NewGomegaWithT(t)
+	ens := setupTestEnsemble(ctx, reconcilertesting.EventTypePrefix, g, natsPort, 7070)
+	defer ens.cancel()
+
+	subscription, subscriptionName := createSubscription(ctx, g, ens,
+		reconcilertesting.WithFilter("", uncleanEventType("")),
+		reconcilertesting.WithSinkURLFromSvc(ens.subscriberSvc),
+	)
+	testSubscriptionOnK8s(ctx, g, ens, subscription, subscriptionName,
+		reconcilertesting.HaveCondition(reconcilertesting.DefaultCondition("")),
+		reconcilertesting.HaveSubscriptionReady(),
+		reconcilertesting.HaveCleanEventTypes([]string{cleanEventType("")}),
+	)
+
+	ens.natsServer.Shutdown()
+	testSubscriptionOnK8s(ctx, g, ens, subscription, subscriptionName, reconcilertesting.HaveSubscriptionNotReady())
+
+	ens.natsServer = startNATS(natsPort)
+	testSubscriptionOnK8s(ctx, g, ens, subscription, subscriptionName, reconcilertesting.HaveSubscriptionReady())
+}
+
+//TestCreateSubscription tests if subscriptions get created properly.
+func TestCreateSubscription(t *testing.T) {
+	ctx := context.Background()
+	g := gomega.NewGomegaWithT(t)
+	ens := setupTestEnsemble(ctx, reconcilertesting.EventTypePrefix, g, 4225, 7075)
+	defer ens.cancel()
+
+	var testCases = []struct {
+		name               string
+		subscriptionOpts   []reconcilertesting.SubscriptionOpt
+		expect             expect
+		shouldTestDeletion bool
+	}{
+		{
+			name: "create and delete",
+			subscriptionOpts: []reconcilertesting.SubscriptionOpt{
+				reconcilertesting.WithFilter(reconcilertesting.EventSource, reconcilertesting.OrderCreatedEventTypeNotClean),
+				reconcilertesting.WithWebhookForNATS(),
+				reconcilertesting.WithSinkURLFromSvc(ens.subscriberSvc),
+			},
+			expect: expect{
+				k8sSubscription: []gomegatypes.GomegaMatcher{
+					reconcilertesting.HaveCondition(reconcilertesting.DefaultCondition("")),
+					reconcilertesting.HaveSubsConfiguration(configDefault(ens.defaultSubscriptionConfig.MaxInFlightMessages)),
+				},
+				natsSubscription: []gomegatypes.GomegaMatcher{
+					reconcilertesting.BeNotNil(),
+					reconcilertesting.BeValid(),
+					reconcilertesting.HaveSubject(reconcilertesting.OrderCreatedEventType),
```

Comment:
nit: out-of-scope because you did not introduce it, but I am interested in your opinion.

If `OrderCreatedEventTypeNotClean ` is the not clean version, wouldn't it be easier to read if `OrderCreatedEventType ` is named `OrderCreatedEventTypeClean` ? 
The reader directly sees: input is **not clean**, output is **clean**. So cleaning worked :)